Allow installing Python bindings with pip

nilsvu opened the pull request:
## Proposed changes

This PR replaces the script that appends the Python modules to the `PYTHONPATH` with functionality that allows installing them with `pip`. The modules can be installed into a Python environment in _development mode_, which means they are symlinked so that changes to the modules will be reflected in the Python environment, with the command `pip install -e path/to/bin/python`.

For those working on Python code for SpECTRE, I would like to note the following (unrelated to this PR):

Once the modules are installed in the Python environment, they can also be run directly from the command line with the command `python -m the_module [args]`. This runs a `__main__.py` script in the package directory or the specific module file as `__main__`. See e.g. https://docs.python.org/3.7/library/__main__.html. We can use this functionality to write Python modules that both provide functions to use e.g. in a Jupyter notebook _and_ expose a CLI, e.g. to make visualisations.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe commented:
I'm fine with adding this functionality, but I think we should keep supporting the `PYTHONPATH` method. Does this PR also support conda by chance? If so, would you mind stating something about that?

nilsvu committed:
Allow installing Python bindings with pip

nilsvu commented:
Ok sounds good, I added the `PYTHONPATH` method back in.

With Anaconda you also install local packages with `pip`, so it works the same way. I added a note to the documentation.

nilsdeppe commented on src/PythonBindings/setup.py:
```diff
@@ -0,0 +1,20 @@
+setup(
+    name='spectre',
+    version='0.1',
+    description="Python bindings for the SpECTRE code",
```
`for the code SpECTRE` or just `for SpECTRE`

nilsvu replied:
Fixed this. I should note that we'll certainly want to extend the metadata in this file in the future, e.g. add a `long_description`, do proper versioning, add contact email addresses, and so on.

nilsdeppe replied:
Ah okay, cool! Once we get away from version 0 I guess we can set that according to the SpECTRE version?

nilsvu replied:
Yeah, we can do that with a CMake replace.

nilsdeppe reviewed (commented):
One very small comment. I like this. Thanks for the conda clarification! That'll be helpful for us python non-experts :)

Edit: I'll explicitly state I'm happy for you to squash this trivial comment :)

kidder commented on src/PythonBindings/setup.py:
```diff
@@ -0,0 +1,20 @@
+
+setup(
+    name='spectre',
+    version='0.1',
```
what does version refer to?  If it is the version of spectre then it should be not be hard-coded.

nilsvu replied:
It is the version of the Python package. I have put no thought into how to version this yet. Is there a CMake variable I can substitute in here? Or do you have another suggestion?

nilsvu replied:
I don’t think we have even decided how to do versioning.. #615

nilsvu replied:
I just substitute in the `SpECTRE_VERSION` here for now. Seems reasonable for the Python bindings to share the main code's version.
